stan/math/rev/core/var.hpp

```diff
@@ -94,6 +95,23 @@ class var_value {
    */
   var_value(vari_type* vi) : vi_(vi) {}  // NOLINT
 
+  /**
+   * Construct a `var_value` with an inner plain matrix type from a `var_value`
+   *  holding an expression.
+   * @tparam S a non-plain eigen expression type
+   * @param other an eigen expression
+   */
+  template <typename S, typename T_ = T,
+            require_convertible_t<S&, value_type>* = nullptr,
+            require_plain_type_t<T_>* = nullptr,
+            require_not_same_t<T, S>* = nullptr>
+  var_value(const var_value<S>& other) : vi_(new vari_type(other.vi_->val_)) {
+    reverse_pass_callback(
+        [this_vi = this->vi_, other_vi = other.vi_]() mutable {
+          other_vi->adj_ += this_vi->adj_;
+        });
+  }
```

Comment:
Should we have each `operator=` have a constructor as well. If feels like it would be good for the API to have the same ones.

Reply:
`require_all_plain_type_t<T, S>* = nullptr` case is handled by defailt copy constructor (T and S can in fact only be same type). We can not define copy assignment operator to do the same as that would also take non-plain types.

`require_any_not_plain_type_t<T, S>* = nullptr` case is handled by this constructor.

We need to explicitely define copy constructor just to delegate to one of the upper two cases, as copy constructor can not be templated and wiould not pich one of those two cases. So I think this is fine.